torchtitan/models/norms.py

```diff
@@ -213,94 +213,147 @@ def _rms_norm_bwd_kernel_sm(
     tl.store(DW + row_block_id * N + cols, dw, mask=mask)
 
 
-class TritonFusedRMSNorm(torch.autograd.Function):
-    @staticmethod
-    def forward(ctx, x, weight, eps):
-        x_shape_start = x.shape
+# Make fused_rmsnorm a custom op, to work around tracing issues for pp tracer/export
+FUSED_RMSNORM_FORWARD = "torchtitan::fused_rmsnorm_forward"
+FUSED_RMSNORM_BACKWARD = "torchtitan::fused_rmsnorm_backward"
+torch.library.define(
+    FUSED_RMSNORM_FORWARD, "(Tensor x, Tensor weight, float eps) -> (Tensor, Tensor)"
+)
```

Comment:
Is the IMA still a problem? If so:

Can you try the following:
```py
torch.library.define(
    FUSED_RMSNORM_FORWARD, "(Tensor x, Tensor weight, float eps) -> (Tensor, Tensor)",
    tags=[torch.Tag.needs_fixed_stride_order])
)
```

Do you know where the IMA is happening? that is, does it happen with torch.compile(backend="eager"), torch.compile(backend="aot_eager"), and/or torch.compile(backend="inductor")?

Reply:
Hi @zou3519  - thanks for the update.
I ran with that modified registered op as above, but hit's an IMA:
~~~
Ima:
rank0]:2024-05-20 12:51:50,341 - root - INFO - Training starts at step 1
[rank0]:[rank0]: Traceback (most recent call last):
[rank0]:[rank0]:   File "/data/users/less/local/torchtitan_oss/train.py", line 420, in <module>
[rank0]:[rank0]:     main(config)
[rank0]:[rank0]:   File "/home/less/local/miniconda3/envs/gocuda/lib/python3.10/site-packages/torch/distributed/elastic/multiprocessing/errors/__init__.py", line 347, in wrapper
[rank0]:[rank0]:     return f(*args, **kwargs)
[rank0]:[rank0]:   File "/data/users/less/local/torchtitan_oss/train.py", line 309, in main
[rank0]:[rank0]:     loss.backward()
[rank0]:[rank0]:   File "/home/less/local/miniconda3/envs/gocuda/lib/python3.10/site-packages/torch/_tensor.py", line 523, in backward
[rank0]:[rank0]:     torch.autograd.backward(
[rank0]:[rank0]:   File "/home/less/local/miniconda3/envs/gocuda/lib/python3.10/site-packages/torch/autograd/__init__.py", line 267, in backward
[rank0]:[rank0]:     _engine_run_backward(
[rank0]:[rank0]:   File "/home/less/local/miniconda3/envs/gocuda/lib/python3.10/site-packages/torch/autograd/graph.py", line 1201, in _engine_run_backward
[rank0]:[rank0]:     return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
[rank0]:[rank0]:   File "/home/less/local/miniconda3/envs/gocuda/lib/python3.10/site-packages/torch/autograd/function.py", line 302, in apply
[rank0]:[rank0]:     return user_fn(self, *args)
[rank0]:[rank0]:   File "/data/users/less/local/torchtitan_oss/torchtitan/models/norms.py", line 360, in backward
[rank0]:[rank0]:     return torch.ops.torchtitan.fused_rmsnorm_backward.default(
[rank0]:[rank0]:   File "/home/less/local/miniconda3/envs/gocuda/lib/python3.10/site-packages/torch/_ops.py", line 630, in __call__
[rank0]:[rank0]:     return self_._op(*args, **kwargs)
[rank0]:[rank0]:   File "/data/users/less/local/torchtitan_oss/torchtitan/models/norms.py", line 328, in fused_rmsnorm_backward
[rank0]:[rank0]:     dw = _dw.sum(0).to(weight.dtype)
[rank0]:[rank0]: RuntimeError: CUDA error: an illegal memory access was encountered
~~~

Reply:
also I'm not sure what you mean regarding torch.compile - this is already a fused kernel so we are running it directly without using torch.compile.
Running regular RMSNorm with torch.compile results in completely different issues (creation of multiple weights or FX graph tracing failures). 
I have made a cuda RMSNorm that I will test out now as a possible replacement for this issue.

Reply:
Sorry, I thought we were encountering the IMA when using this custom op with torch.compile. It is the case that the custom op IMAs in eager mode PyTorch?